File: benchmarks/mandelbrot/zig/mandelbrot.zig
//! Mandelbrot: each pixel of an N x N grid maps onto a fixed viewport of the
//! complex plane and iterates z <- z^2 + c until |z| > 2 or max_iter is reached.
//!
//! The program prints two integers on stdout: the checksum (the sum of every
//! pixel's iteration count) and the wall-clock nanoseconds spent computing it.
//!
//! THE CHECKSUM IS THE HARNESS'S CORRECTNESS GATE. In strict floating-point mode
//! it must be bit-identical across every compiler, every language and both ISAs.
//! That only holds because the operations below are multiply, add, subtract and
//! compare, all correctly rounded by IEEE 754 -- and because every implementation
//! evaluates them in exactly this order. Do not "simplify" the arithmetic: any
//! reassociation, or an FMA contraction of `zr2 - zi2 + cr`, changes the last
//! bit, flips a boundary pixel from 999 to 1000 iterations, and breaks the
//! invariant. See METHODOLOGY.md#the-strict-mode-invariant.
//!
//! Zig's float mode is `.strict` by default and this kernel leaves it there.
//! Relaxing it is `@setFloatMode(.optimized)` -- a statement *in the source*, not
//! a flag on the compiler -- so `fma` and `fast` cannot be build args over one
//! kernel the way they are for C. A kernel with a different float mode is a
//! different kernel, and this backend declares `strict` alone rather than pretend
//! otherwise. See `bench.yaml`.

const std = @import("std");

/// The viewport. Part of the cross-implementation contract: changing any of these
/// constants changes the reference checksum.
const X_MIN: f64 = -2.0;
const X_MAX: f64 = 0.5;
const Y_MIN: f64 = -1.25;
const Y_MAX: f64 = 1.25;

/// Sum the iteration counts of one row. The unit of work.
///
/// With any realistic n there are far more rows than threads, which is what the
/// dynamic hand-out below needs: the load is imbalanced by design (interior pixels
/// run to max_iter, exterior ones exit after a few iterations), so a static
/// contiguous split would measure the split rather than the backend.
fn rowIterations(row: u32, n: u32, max_iter: u32, dx: f64, dy: f64) u64 {
    const ci = Y_MIN + (@as(f64, @floatFromInt(row)) + 0.5) * dy;
    var sum: u64 = 0;

    var col: u32 = 0;
    while (col < n) : (col += 1) {
        const cr = X_MIN + (@as(f64, @floatFromInt(col)) + 0.5) * dx;
        var zr: f64 = 0.0;
        var zi: f64 = 0.0;
        var iter: u32 = 0;

        while (iter < max_iter) {
            const zr2 = zr * zr;
            const zi2 = zi * zi;
            if (zr2 + zi2 > 4.0) {
                break;
            }
            zi = 2.0 * zr * zi + ci;
            zr = zr2 - zi2 + cr;
            iter += 1;
        }
        sum += iter;
    }
    return sum;
}

/// One worker: takes the next row until there are none left, then parks its
/// subtotal where `main` can find it. No mutex -- the atomic cursor is the only
/// thing the workers share.
const Worker = struct {
    n: u32,
    max_iter: u32,
    dx: f64,
    dy: f64,
    next_row: *std.atomic.Value(u32),
    sum: u64 = 0,

    fn run(self: *Worker) void {
        var sum: u64 = 0;
        while (true) {
            const row = self.next_row.fetchAdd(1, .monotonic);
            if (row >= self.n) {
                break;
            }
            sum += rowIterations(row, self.n, self.max_iter, self.dx, self.dy);
        }
        self.sum = sum;
    }
};

fn parsePositive(text: []const u8, name: []const u8) u32 {
    const value = std.fmt.parseInt(u32, text, 10) catch {
        std.debug.print("{s} must be a positive integer, got `{s}`\n", .{ name, text });
        std.process.exit(2);
    };
    if (value == 0) {
        std.debug.print("{s} must be a positive integer, got `{s}`\n", .{ name, text });
        std.process.exit(2);
    }
    return value;
}

pub fn main() !void {
    const allocator = std.heap.smp_allocator;

    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    if (args.len != 4) {
        std.debug.print("usage: {s} <n> <max_iter> <threads>\n", .{args[0]});
        std.process.exit(2);
    }

    // Never comptime constants: a backend would fold the whole computation away
    // and the benchmark would measure nothing, very quickly.
    const n = parsePositive(args[1], "n");
    const max_iter = parsePositive(args[2], "max_iter");
    const threads = parsePositive(args[3], "threads");

    const dx = (X_MAX - X_MIN) / @as(f64, @floatFromInt(n));
    const dy = (Y_MAX - Y_MIN) / @as(f64, @floatFromInt(n));

    var next_row = std.atomic.Value(u32).init(0);
    const workers = try allocator.alloc(Worker, threads);
    defer allocator.free(workers);
    const handles = try allocator.alloc(std.Thread, threads);
    defer allocator.free(handles);

    // Thread creation is inside the timer on purpose: spawning the pool is part
    // of what a parallel runtime costs, and the point is to compare runtimes.
    var timer = try std.time.Timer.start();

    for (workers, handles) |*worker, *handle| {
        worker.* = .{
            .n = n,
            .max_iter = max_iter,
            .dx = dx,
            .dy = dy,
            .next_row = &next_row,
        };
        handle.* = try std.Thread.spawn(.{}, Worker.run, .{worker});
    }
    for (handles) |handle| {
        handle.join();
    }
    const elapsed_ns = timer.read();

    // Summing 64-bit integers is associative, so the reduction order cannot
    // perturb the checksum however the threads happened to finish.
    var checksum: u64 = 0;
    for (workers) |worker| {
        checksum += worker.sum;
    }

    // Printing the checksum is what stops dead-code elimination from deleting the
    // loop above.
    var line: [64]u8 = undefined;
    const printed = try std.fmt.bufPrint(&line, "{d} {d}\n", .{ checksum, elapsed_ns });
    try std.fs.File.stdout().writeAll(printed);
}

File: benchmarks/mandelbrot/julia/mandelbrot.jl
# Mandelbrot: each pixel of an N x N grid maps onto a fixed viewport of the
# complex plane and iterates z <- z^2 + c until |z| > 2 or max_iter is reached.
#
# The program prints two integers on stdout: the checksum (the sum of every
# pixel's iteration count) and the wall-clock nanoseconds spent computing it.
#
# THE CHECKSUM IS THE HARNESS'S CORRECTNESS GATE. In strict floating-point mode it
# must be bit-identical to every other implementation's, C included. That holds
# because Julia's Float64 is an IEEE 754 double, because multiply, add, subtract
# and compare are correctly rounded, and because Julia never contracts a
# multiply-add into an FMA behind your back -- fusing is `muladd`, which you have
# to write. It also requires the arithmetic below to be evaluated in exactly the
# order the C kernel uses: do not "simplify" it.
# See METHODOLOGY.md#the-strict-mode-invariant.
#
# Julia compiles -- this kernel runs as native code, through LLVM, exactly like
# the C one -- but it compiles *during the run*, on first call, which is why the
# manifest declares an interpreter and no compiler: nothing is compiled ahead of
# the run. That JIT pause sits inside the timer, and it is the result this row
# exists to publish.
#
# No third-party dependency, by design: a package would drag in a load path and
# inflate the very startup this benchmark measures.

# The viewport. Part of the cross-implementation contract: changing any of these
# constants changes the reference checksum. `const`, because a non-constant global
# is boxed and type-unstable, and the hot loop would be reading `Any`.
const X_MIN = -2.0
const X_MAX = 0.5
const Y_MIN = -1.25
const Y_MAX = 1.25

"""
Sum the iteration counts of one row. The unit of work.

With any realistic `n` there are far more rows than threads, which is what the
dynamic hand-out below needs: the load is imbalanced by design (interior pixels
run to `max_iter`, exterior ones exit after a few iterations), so a static
contiguous split would measure the split rather than the backend.
"""
function row_iterations(row::Int, n::Int, max_iter::Int, dx::Float64, dy::Float64)::Int
    ci = Y_MIN + (row + 0.5) * dy
    total = 0

    for col in 0:(n - 1)
        cr = X_MIN + (col + 0.5) * dx
        zr = 0.0
        zi = 0.0
        iter = 0

        while iter < max_iter
            zr2 = zr * zr
            zi2 = zi * zi
            if zr2 + zi2 > 4.0
                break
            end
            zi = 2.0 * zr * zi + ci
            zr = zr2 - zi2 + cr
            iter += 1
        end
        total += iter
    end
    return total
end

"""
One worker's whole life: take the next row until there are none left.

`Threads.atomic_add!` returns the value *before* the addition, which is this
language's `fetch_add`: the cursor is the only state the workers share.
"""
function work(cursor::Threads.Atomic{Int}, n::Int, max_iter::Int, dx::Float64, dy::Float64)::Int
    total = 0
    while true
        row = Threads.atomic_add!(cursor, 1)
        if row >= n
            break
        end
        total += row_iterations(row, n, max_iter, dx, dy)
    end
    return total
end

function parse_positive(text::AbstractString, name::AbstractString)::Int
    value = tryparse(Int, text)
    if value === nothing || value <= 0
        println(stderr, "$name must be a positive integer, got `$text`")
        exit(2)
    end
    return value
end

function main()
    if length(ARGS) != 3
        println(stderr, "usage: mandelbrot.jl <n> <max_iter> <threads>")
        exit(2)
    end

    # Never module-level constants: a backend could fold the computation away.
    n = parse_positive(ARGS[1], "n")
    max_iter = parse_positive(ARGS[2], "max_iter")
    threads = parse_positive(ARGS[3], "threads")

    dx = (X_MAX - X_MIN) / n
    dy = (Y_MAX - Y_MIN) / n
    cursor = Threads.Atomic{Int}(0)

    # Spawning the tasks is inside the timer on purpose: what a parallel runtime
    # costs to start is part of what that runtime costs. So, here, is the JIT: the
    # first call to `row_iterations` compiles it, and that pause is a property of
    # this backend rather than an artefact to be warmed away.
    #
    # The kernel never asks how many threads the machine has: it spawns exactly as
    # many tasks as the harness asked for. `Threads.nthreads()` would answer with
    # whatever `--threads` the entrypoint passed, which is the same number by a
    # longer route -- and would quietly disagree with the harness the day a runtime
    # reads the cgroup quota differently.
    started = time_ns()
    tasks = [Threads.@spawn work(cursor, n, max_iter, dx, dy) for _ in 1:threads]
    # Summing 64-bit integers is associative, so the order in which the workers
    # finish cannot perturb the checksum.
    checksum = sum(fetch(task)::Int for task in tasks)
    elapsed_ns = time_ns() - started

    # Printing the checksum is what stops the compiler from eliding the loop above.
    println("$checksum $elapsed_ns")
    return 0
end

# Only when run as a program: the timed build phase `include`s this file to
# compile it, and must not execute the workload while doing so.
if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
